### ISSUE TITLE
Use :binary_id for correlation_id and causation_id

### ISSUE THREAD

**Issue by DominikGuzei** [slashdotdash/eventstore#51]
Is there any reason why these two id fields are saved as type `text` in Postgres?
It's not the end of the world but causing me issues because im using these id's as primary keys in ecto schemas and it feels weird to use `:string` there when i know that these are UUIDs.

**Comment by slashdotdash** [comment]
Seems like a reasonable change to make, then we can ensure these fields are always UUIDs. Don't see why that would be a problem.

**Comment by DominikGuzei** [comment]
Ok sounds great, i just wondered if there was some deeper thought behind the `text` type (maybe arbitrary, more complex identifiers?) but i don't think that's necessary or common practice 😉

**Comment by slashdotdash** [comment]
@DominikGuzei Feel free to have a go at making this change, you might find it only requires changing the SQL statement to create the `events` table.

**Comment by DominikGuzei** [comment]
@slashdotdash PR is up now … sorry for the long delay, was on holidays and occupied by other projects 😉 the change to `uuid` was easy but im running into another issue now when i define my projection schema like this:

```elixir
defmodule Josoor.Server.TranslationProjection do
  
  use Josoor.Server.Projection
  
  @primary_key {:translation_uuid, :binary_id, []}
  
  schema "translations" do
    field :content_uuid, :binary_id
    field :language, :string
    field :content, :string
    timestamps()
  end

end
```

something is wrong with the encoding:
```shell
[error] GenServer #PID<0.886.0> terminating
** (Postgrex.Error) ERROR 22021 (character_not_in_repertoire): invalid byte sequence for encoding "UTF8": 0xc1 0x7b
    (ecto) lib/ecto/adapters/sql.ex:571: Ecto.Adapters.SQL.struct/7
    (ecto) lib/ecto/repo/schema.ex:469: Ecto.Repo.Schema.apply/4
    (ecto) lib/ecto/repo/schema.ex:205: anonymous fn/13 in Ecto.Repo.Schema.do_insert/4
    (ecto) lib/ecto/multi.ex:422: Ecto.Multi.apply_operation/5
    (elixir) lib/enum.ex:1811: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto) lib/ecto/multi.ex:412: anonymous fn/5 in Ecto.Multi.apply_operations/5
    (ecto) lib/ecto/adapters/sql.ex:620: anonymous fn/3 in Ecto.Adapters.SQL.do_transaction/3
    (db_connection) lib/db_connection.ex:1275: DBConnection.transaction_run/4
    (db_connection) lib/db_connection.ex:1199: DBConnection.run_begin/3
    (db_connection) lib/db_connection.ex:790: DBConnection.transaction/3
    (ecto) lib/ecto/repo/queryable.ex:21: Ecto.Repo.Queryable.transaction/4
    (server) lib/server/projections/translation_projection.ex:3: Josoor.Server.TranslationProjection.update_projection/3
    (commanded) lib/commanded/event/handler.ex:101: anonymous fn/2 in Commanded.Event.Handler.handle_info/2
    (elixir) lib/enum.ex:1811: Enum."-reduce/3-lists^foldl/2-0-"/3
    (commanded) lib/commanded/event/handler.ex:96: Commanded.Event.Handler.handle_info/2
    (stdlib) gen_server.erl:616: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:686: :gen_server.handle_msg/6
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
Last message: {:events, [%Commanded.EventStore.RecordedEvent{causation_id: nil, correlation_id: "9ca76c9e-b46d-4d9c-b01e-3364bf1ffc50", created_at: ~N[2017-08-06 20:19:01.827036], data: %Josoor.Domain.Translation.Provided{content: "test translation content", content_uuid: "54c17b22-516c-4f70-a488-b5307e7becfa", language: "eng", translation_uuid: "0e6ebafc-d409-4763-9271-ebddaeb0298c"}, event_number: 14, event_type: "Elixir.Josoor.Domain.Translation.Provided", metadata: %{}, stream_id: 6, stream_version: 1}]}
State: %Commanded.Event.Handler{handler_module: Josoor.Server.TranslationsProjector, handler_name: "Josoor.Server.TranslationProjection", last_seen_event: nil, subscribe_from: :origin, subscription: #PID<0.887.0>}
```

**Comment by slashdotdash** [comment]
Can you include the code that is populating the Ecto schema?

**Comment by slashdotdash** [comment]
```console
iex(1)> uuid = "9ca76c9e-b46d-4d9c-b01e-3364bf1ffc50"
"9ca76c9e-b46d-4d9c-b01e-3364bf1ffc50"

iex(2)> UUID.info!(uuid)
[uuid: "9ca76c9e-b46d-4d9c-b01e-3364bf1ffc50",
 binary: <<156, 167, 108, 158, 180, 109, 77, 156, 176, 30, 51, 100, 191, 31,
   252, 80>>, type: :default, version: 4, variant: :rfc4122]

iex(3)> UUID.string_to_binary!(uuid)
<<156, 167, 108, 158, 180, 109, 77, 156, 176, 30, 51, 100, 191, 31, 252, 80>>
```

Try using the `UUID.string_to_binary!/1` function to convert the UUID (as a string) from the event into your Ecto field.

**Comment by DominikGuzei** [comment]
Yes i migrated the events table and it worked without any problem … but i also figured that im not really using the `correlation_id` anywhere yet so that cannot be related. My problem seems to to stem from encoding the UUIDs with poison and then back.

Btw. im using `Ecto.UUID` for all UUIDs now … is that a problem?

When i convert the uuids to binary at the insert
```elixir
Ecto.Multi.insert(multi, :translations, %TranslationProjection{
        translation_uuid: UUID.string_to_binary!(translation_uuid),
        content_uuid: UUID.string_to_binary!(content_uuid),
        language: language,
        content: content,
      })
```
im getting this error:
```bash
[error] GenServer #PID<0.895.0> terminating
** (Ecto.ChangeError) value `<<84, 193, 123, 34, 81, 108, 79, 112, 164, 136, 181, 48, 126, 123, 236, 250>>` for `Josoor.Server.TranslationProjection.content_uuid` in `insert` does not match type :binary_id
    (ecto) lib/ecto/repo/schema.ex:706: Ecto.Repo.Schema.dump_field!/6
    (ecto) lib/ecto/repo/schema.ex:715: anonymous fn/6 in Ecto.Repo.Schema.dump_fields!/5
    (stdlib) lists.erl:1263: :lists.foldl/3
    (ecto) lib/ecto/repo/schema.ex:713: Ecto.Repo.Schema.dump_fields!/5
    (ecto) lib/ecto/repo/schema.ex:662: Ecto.Repo.Schema.dump_changes!/6
    (ecto) lib/ecto/repo/schema.ex:200: anonymous fn/13 in Ecto.Repo.Schema.do_insert/4
    (ecto) lib/ecto/multi.ex:422: Ecto.Multi.apply_operation/5
    (elixir) lib/enum.ex:1811: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto) lib/ecto/multi.ex:412: anonymous fn/5 in Ecto.Multi.apply_operations/5
    (ecto) lib/ecto/adapters/sql.ex:620: anonymous fn/3 in Ecto.Adapters.SQL.do_transaction/3
    (db_connection) lib/db_connection.ex:1275: DBConnection.transaction_run/4
    (db_connection) lib/db_connection.ex:1199: DBConnection.run_begin/3
    (db_connection) lib/db_connection.ex:790: DBConnection.transaction/3
    (ecto) lib/ecto/repo/queryable.ex:21: Ecto.Repo.Queryable.transaction/4
    (server) lib/server/projections/translation_projection.ex:3: Josoor.Server.TranslationProjection.update_projection/3
    (commanded) lib/commanded/event/handler.ex:101: anonymous fn/2 in Commanded.Event.Handler.handle_info/2
    (elixir) lib/enum.ex:1811: Enum."-reduce/3-lists^foldl/2-0-"/3
    (commanded) lib/commanded/event/handler.ex:96: Commanded.Event.Handler.handle_info/2
    (stdlib) gen_server.erl:616: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:686: :gen_server.handle_msg/6
Last message: {:events, [%Commanded.EventStore.RecordedEvent{causation_id: nil, correlation_id: "9ca76c9e-b46d-4d9c-b01e-3364bf1ffc50", created_at: ~N[2017-08-06 20:19:01.827036], data: %Josoor.Domain.Translation.Provided{content: "test translation content", content_uuid: "54c17b22-516c-4f70-a488-b5307e7becfa", language: "eng", translation_uuid: "0e6ebafc-d409-4763-9271-ebddaeb0298c"}, event_number: 14, event_type: "Elixir.Josoor.Domain.Translation.Provided", metadata: %{}, stream_id: 6, stream_version: 1}]}
State: %Commanded.Event.Handler{handler_module: Josoor.Server.TranslationsProjector, handler_name: "Josoor.Server.TranslationProjection", last_seen_event: nil, subscribe_from: :origin, subscription: #PID<0.896.0>}
```

**Comment by slashdotdash** [comment]
Sorry, I think you need to use:

```elixir
{:ok, decoded} = Ecto.UUID.dump(uuid)
```

**Comment by DominikGuzei** [comment]
That throws the same error as with `UUID.string_to_binary!` as it does exactly the same …

**Comment by slashdotdash** [comment]
I've just taken a look at the example Phoenix app I'm building with Commanded and it has a [projection that is copying a UUID string from an event](https://github.com/slashdotdash/conduit/blob/382567f35e7eb9e94d3ce5fbde5368390d35b916/lib/conduit/blog/projectors/article.ex#L86):

```elixir
defmodule Conduit.Blog.Projectors.Article do
  use Commanded.Projections.Ecto, name: "Blog.Projectors.Article"

  project %ArticleCommented{} = commented, %{created_at: commented_at} do
    multi
    |> Ecto.Multi.run(:author, fn _changes -> get_author(commented.author_uuid) end)
    |> Ecto.Multi.run(:comment, fn %{author: author} ->
      comment = %Comment{
        uuid: commented.uuid,
        body: commented.body,
        article_uuid: commented.article_uuid,
        author_uuid: author.uuid,
        author_username: author.username,
        author_bio: author.bio,
        author_image: author.image,
        commented_at: commented_at,
      }

      Repo.insert(comment)
    end)
  end
end
```

Here's the [Comment schema](https://github.com/slashdotdash/conduit/blob/382567f35e7eb9e94d3ce5fbde5368390d35b916/lib/conduit/blog/comment.ex):

```elixir
defmodule Conduit.Blog.Comment do
  use Ecto.Schema

  @primary_key {:uuid, :binary_id, autogenerate: false}

  schema "blog_comments" do
    field :body, :string
    field :article_uuid, :binary_id
    field :author_uuid, :binary_id
    field :author_bio, :string
    field :author_image, :string
    field :author_username, :string
    field :commented_at, :naive_datetime
    
    timestamps()
  end
end
```

**Comment by DominikGuzei** [comment]
Hmm ok so the only difference i see is how you define the projector. Here is mine:
```elixir
defmodule Josoor.Server.TranslationsProjector do
  
  import Destructure
  
  alias Josoor.Server.TranslationProjection
  alias Josoor.Domain.Translation.Provided
  
  @behaviour Commanded.Event.Handler
  @projection_name "translations"

  def handle(
    d(%Provided{translation_uuid, content_uuid, language, content}),
    d(%{event_number})
  ) do
    TranslationProjection.update_projection(@projection_name, event_number, fn multi ->
      Ecto.Multi.insert(multi, :translations, %TranslationProjection{
        translation_uuid: Ecto.UUID.dump(translation_uuid),
        content_uuid: Ecto.UUID.dump(content_uuid),
        language: language,
        content: content,
      })
    end)
  end

  # ignore all other events
  def handle(_event, _metadata), do: :ok
end
```
i kind of pieced this together from various examples and articles … do you see anything wrong with it?

**Comment by slashdotdash** [comment]
Remove the `Ecto.UUID.dump/1` function calls, just try setting the values directly (from the event fields):

```elixir
TranslationProjection.update_projection(@projection_name, event_number, fn multi ->
  Ecto.Multi.insert(multi, :translations, %TranslationProjection{
    translation_uuid: translation_uuid,
    content_uuid: content_uuid,
    language: language,
    content: content,
  })
end)
```

**Comment by DominikGuzei** [comment]
This gives me the error i posted first:
```bash
[error] GenServer #PID<0.895.0> terminating
** (Ecto.ChangeError) value `<<84, 193, 123, 34, 81, 108, 79, 112, 164, 136, 181, 48, 126, 123, 236, 250>>` for `Josoor.Server.TranslationProjection.content_uuid` in `insert` does not match type :binary_id
    (ecto) lib/ecto/repo/schema.ex:706: Ecto.Repo.Schema.dump_field!/6
    (ecto) lib/ecto/repo/schema.ex:715: anonymous fn/6 in Ecto.Repo.Schema.dump_fields!/5
    (stdlib) lists.erl:1263: :lists.foldl/3
    (ecto) lib/ecto/repo/schema.ex:713: Ecto.Repo.Schema.dump_fields!/5
    (ecto) lib/ecto/repo/schema.ex:662: Ecto.Repo.Schema.dump_changes!/6
    (ecto) lib/ecto/repo/schema.ex:200: anonymous fn/13 in Ecto.Repo.Schema.do_insert/4
    (ecto) lib/ecto/multi.ex:422: Ecto.Multi.apply_operation/5
    (elixir) lib/enum.ex:1811: Enum."-reduce/3-lists^foldl/2-0-"/3
    (ecto) lib/ecto/multi.ex:412: anonymous fn/5 in Ecto.Multi.apply_operations/5
    (ecto) lib/ecto/adapters/sql.ex:620: anonymous fn/3 in Ecto.Adapters.SQL.do_transaction/3
    (db_connection) lib/db_connection.ex:1275: DBConnection.transaction_run/4
    (db_connection) lib/db_connection.ex:1199: DBConnection.run_begin/3
    (db_connection) lib/db_connection.ex:790: DBConnection.transaction/3
    (ecto) lib/ecto/repo/queryable.ex:21: Ecto.Repo.Queryable.transaction/4
    (server) lib/server/projections/translation_projection.ex:3: Josoor.Server.TranslationProjection.update_projection/3
    (commanded) lib/commanded/event/handler.ex:101: anonymous fn/2 in Commanded.Event.Handler.handle_info/2
    (elixir) lib/enum.ex:1811: Enum."-reduce/3-lists^foldl/2-0-"/3
    (commanded) lib/commanded/event/handler.ex:96: Commanded.Event.Handler.handle_info/2
    (stdlib) gen_server.erl:616: :gen_server.try_dispatch/4
    (stdlib) gen_server.erl:686: :gen_server.handle_msg/6
Last message: {:events, [%Commanded.EventStore.RecordedEvent{causation_id: nil, correlation_id: "9ca76c9e-b46d-4d9c-b01e-3364bf1ffc50", created_at: ~N[2017-08-06 20:19:01.827036], data: %Josoor.Domain.Translation.Provided{content: "test translation content", content_uuid: "54c17b22-516c-4f70-a488-b5307e7becfa", language: "eng", translation_uuid: "0e6ebafc-d409-4763-9271-ebddaeb0298c"}, event_number: 14, event_type: "Elixir.Josoor.Domain.Translation.Provided", metadata: %{}, stream_id: 6, stream_version: 1}]}
State: %Commanded.Event.Handler{handler_module: Josoor.Server.TranslationsProjector, handler_name: "Josoor.Server.TranslationProjection", last_seen_event: nil, subscribe_from: :origin, subscription: #PID<0.896.0>}
```

**Comment by DominikGuzei** [comment]
Maybe the problem is the way i validate the commands?

```elixir
defmodule Josoor.Domain.Translation.Provide do
  use Vex.Struct
  defstruct [
    :translation_uuid,
    :content_uuid,
    :language,
    :content
  ]
  validates :translation_uuid, &(Ecto.UUID.cast!(&1))
  validates :content_uuid, &(Ecto.UUID.cast!(&1))
  validates :language, presence: true
  validates :content, presence: true
end
```

**Comment by DominikGuzei** [comment]
Ok no … i tried it with `UUID.info!` and it's the same result.

**Comment by slashdotdash** [comment]
They're valid UUIDs, so that's not causing the issue. I'm running out of ideas. 

Check you have latest mix dependencies (`mix hex.outdated`)?
Rebuild database (`mix do event_store.drop, event_store.create, ecto.reset`)?

**Comment by slashdotdash** [comment]
You could try running the insert via an `iex -S mix` console using copy/paste values, see if it still errors then:

```elixir
Repo.insert!(%TranslationProjection{
  translation_uuid: "0e6ebafc-d409-4763-9271-ebddaeb0298c",
  content_uuid: "54c17b22-516c-4f70-a488-b5307e7becfa",
  language: "eng",
  content: "test translation content",
})

**Comment by DominikGuzei** [comment]
Yeah even if i use copy paste values in the insert logic it fails with same error 😞

**Comment by DominikGuzei** [comment]
Im using the latest packages for everything (all green with `mix hex.outdated`).
I also resetted all DBs etc. multiple times already but that issue remains.

**Comment by slashdotdash** [comment]
Sorry, I'm finishing up for the night now. If you don't make progress, send me a copy/gist of the schema & database migration to look at tomorrow.

**Comment by DominikGuzei** [comment]
Yeah me too, btw. i wrote you on elixir-lang slack … might be better for such detailed debugging sessions 😉 thanks a lot for the support so far!

**Comment by DominikGuzei** [comment]
OMG … it was totally my fault, i forgot to update the migration for the `translations` table. Everything works now! 🎉 sorry for the hassle!

**Comment by slashdotdash** [comment]
Glad to hear you're sorted.

**Comment by slashdotdash** [comment]
Fixed by #79.
